### DISCUSSION TITLE
JobRunr 4.0.0-RC1

### DISCUSSION THREAD

**Discussion by rdehuyss** [jobrunr/jobrunr#214] [General]
Hi all,

JobRunr 4.0.0-RC1 is available in Maven Central and I'm looking for feedback in this thread. Drop me a note here if you are having troubles, ... 

I will not release JobRunr 4.0.0 until a few people confirmed that JobRunr 4.0.0-RC1 works fine for them.

Tagging @tan9 @neildeng @mendess @JoonasC

Cheers!
R

**Comment by tan9** [comment]
@rdehuyss  I can confirm that 4.0.0-RC1 is working as except in my application, including the normal job scheduling/running, `@Recurring` registration, and the Metric exporting. :+1:

**Reply by tan9** [reply]
And I found a typo in https://github.com/jobrunr/jobrunr/releases/tag/v4.0.0-RC1

  - Support for multiple datasources in Spring (a specific datasource for **JobRun**)

**Reply by rdehuyss** [reply]
Thx @tan9 ! Did you also try the new `JobSchedulerRequest`? I'm also looking for feedback on that...

Thanks again!

**Reply by tan9** [reply]
I can not find the context about the `JobRequet`-related designs, and the commit e9e0c4e6736824d9a3ca4dba18980087882d27b5 reference to #210 seems unrelated to this addition.

As of my acknowledgment, is it a workaround for #183? And it looks like merely a Command Pattern design to encapsulate the job method, parameters, and the `@Job` annotation not to be annotated on the jobRequestHandler's `run` method.

To be honest, I don't think this is a good addition to JobRunr:
1. It doubled the complexities of JobRunr's public API.
2. People can write an indirect method (and can be annotated by `@Job`) to achieve the same goal as implementing a `JobReuqest`.
3. I think people will continue to seek a method to access the `JobDetails`. Because they neither want to wrap their job in another method nor write another JobRequest instead, they just want to let JobRunr know the actual method they want to schedule!

**Reply by tan9** [reply]
To summarize, I don't think `JobRequestScheudler` is a good idea unless it can provide direct access to an existing method to be scheduled.

To achieve the goal of direct access to an existing method. Is it practical to use Java 8's method reference and generic for job scheduling? Just like:

```java
jobRequestScheudler.enqueue(System.out::println, "my awesome job!");
```

or

```java
jobScheduler.enqueue(JobRequest.of(System.out::println, "my awesome job!"));
```

If this is practical, maybe we can provide the same functional interfaces as `https://docs.vavr.io/#_functions` provides, like `CheckedFunction1`, `CheckedFunction2`,... `CheckedFunction8`.

**Reply by rdehuyss** [reply]
Hi @tan9 , thanks for your feedback, really appreciate it.

Regarding the commits for `JobRequest` and `JobRequestScheduler` - I indeed did not do that well, I perhaps even had better created a different branch for the `JobRequest` and `JobRequestScheduler` changes.

Why did I add them? Mainly two reasons:
- there are folks that are hesitant to use JobRunr as they do not like the magic of the job analysis. They see it accepts an interface, pass an implementation of that interface (so a class with the `run` method implemented) and JobRunr off-course doesn't like that. See also the first comments here for more info: https://www.reddit.com/r/java/comments/hweabk/v0912_of_jobrunr_a_distributed_job_scheduling/fzhy86e/?context=3
- then there is the GraalVM native stuff - JobRunr cannot (at least currently) support GraalVM native mode as the `SerializedLambda` is not available in native mode. By passing the `JobRequest` all of this magic is bypassed and JobRunr can work on GraalVM native mode.

I agree the public API complicates a bit but that's also the reason I separated the new `JobRequest` methods in the new `JobRequestScheduler` class and did not add then to the existing `JobScheduler` - it's up to the end user/developer to choose whatever API he likes best and if you keep injecting the existing `JobScheduler` you don't even notice the new API.

Does this make sense?

**Reply by tan9** [reply]
Hi @rdehuyss

I agree with you for the job analysis and GraalVM native stuff part. But I think the JobRequest interface seems not as clear as other parts of JobRunr.

I don't know how to implement the interface properly without taking a look at the `TestJobRequest` and `TestJobRequestHandler`:
1. `JobRequest` is a uber class that provides the information of the handler, provides the job metadata, and holds the parameter itself. While Job execution, then it will play as a container of the actual parameter for the job execution. I think it can be neater by simply providing all required parts for Job itself, including target method, parameter, and metadata. That is a `JobDetails` with `@Job`!
2. What if there is a `@Job` annotation on `jobReuqestHandler.run()`? `JobRequest.getJobDisplayName()` and `JobRequest.getMaxRetryCount()` isn't referenced in current codebase. And if we try to really implement it, the `DisplayNameFilter` and `RetryFilter` must take extra effort to get the metadata.  
Furthermore, what will happen if a user enqueue a job that way: 
    ```java
    enqueue(testJobRequestHandler -> testJobRequestHandler.run(new TestJobRequest("inputValue")))
    ```
    and annotate `testJobRequestHandler.run` with `@Job` specifying value different from values from `TestJobRequest`'s properties?
3. `TestJobRequestHandler.jobContext` is another "magic" to me and to the IoC container.

**Reply by tan9** [reply]
I tried the method reference thing, it is too easy to encounter ambiguous methods.

But how about just provide something like this:
```java
    public <T extends JobRequest> JobId enqueue(Class<JobRequestHandler<T>> handlerClass, T request);
```

It is almost equivalent to the current JobRequest <-> JobRequestHanlder relation, and even more simple and descriptive!
People can still annotate their own JobRequestHandler imlementation's `run` method with already well-established `@Job`, let `T` simply as the container of parameters, or even simpler, the value itself if we changed the `extends JobRequest` part to `extends Serializable`.

**Reply by tan9** [reply]
For the `JobContext` thing, maybe a Spring-styled `JobContextAware` interface with a `ThreadLocal`-based `JobContext` proxy is enough:

**Reply by rdehuyss** [reply]
Hi @tan9 ,

thanks again for your great input. I will do it as follows:
- the `JobRequest` interface will only have one method: `getJobRequestHandler()` (so not anymore the methods `getJobDisplayName()` and `getMaxRetryCount()`)
- to set the job display name and retry count, people should annotate the `run` method with the `@Job` annotation
- I really like your idea about the `JobContextAware` - I'm happy I had your second pair of eyes looking into it

Regarding this approach:
```java
    public <T extends JobRequest> JobId enqueue(Class<JobRequestHandler<T>> handlerClass, T request);
```
I personally am not a fan about it as it requires a lot of boilerplate each time you create a job. Now, you can hardcode it once in the class implementing the `JobRequest` and you're done.

I think the above brings together the best of both our idea's.

**Reply by tan9** [reply]
@rdehuyss  I was convinced by you. The only thing left is to educate users with this design pattern :)

Besides, I noticed that `JobContext` injection seems not thread-safe, for Spring ApplicationContext, all `JobRequestHandler` instances are singletons by default. If there are multiple `JobRequest` running on a host, we may be in trouble.

**Reply by rdehuyss** [reply]
Hi @tan9 , I was unable to get it to work in a thread-safe manner with the `JobContextAware` interface. The moment you put a field on a spring bean, it is not thread safe anymore.

I did manage it by adding a default method on the `JobRequestHandler`. 

What do you think of this approach? If this is ok, I will create another RC.

**Reply by tan9** [reply]
@rdehuyss First thought that came into my head is the `@Autowired HttpServletRequest` in Spring-managed beans, a proxy implementation backed by `ThreadLocal`-bounded `HttpServletRequest` instance. Is the default method is implemented by something like this?

**Reply by tan9** [reply]
@rdehuyss  I saw the commit 801420b3e5a953a3af963427c10551a04e4812d3. And it is in at the `BackgroundJobRunner` level, everyone has a  second chance to access the JobContext for the free :+1:

**Reply by tan9** [reply]
One more thing. Put the `ThreadLocalJobContext.clear();` statement in a `finally` block will reduce the change of JobContext leakage.
